# Ejemplar: Realizar ingeniería de características


## **Introducción**


Mientras aprendes, los profesionales de datos que trabajan en proyectos de modelado utilizan ingeniería de características para ayudarlos a determinar qué atributos en los datos pueden predecir mejor ciertas medidas.

En esta actividad, estás trabajando para una empresa que proporciona información a la Asociación Nacional de Baloncesto (NBA), una liga profesional de baloncesto en Norteamérica. Ayudarás a los gerentes y entrenadores de la NBA a identificar qué jugadores tienen más probabilidades de prosperar en el entorno de alta presión del baloncesto profesional y ayudar al equipo a tener éxito con el tiempo.

Para hacer esto, analizarás un subconjunto de datos que contiene información sobre jugadores de la NBA y sus registros de rendimiento. Realizarás ingeniería de características para determinar qué características predicen de manera más efectiva si la carrera de un jugador en la NBA durará al menos cinco años. Los conocimientos obtenidos luego se utilizarán en la siguiente etapa del proyecto: construir el modelo predictivo.


## **Paso 1: Importaciones**


Comienza importando `pandas`.


In [1]:
# Import pandas.

### YOUR CODE HERE ###

import pandas as pd

El conjunto de datos es un archivo .csv llamado `nba-players.csv`. Consiste en registros de rendimiento para un subconjunto de jugadores de la NBA. Como se muestra en esta celda, el conjunto de datos ha sido cargado automáticamente para ti. No necesitas descargar el archivo .csv, ni proporcionar más código, para acceder al conjunto de datos y continuar con este laboratorio. Por favor, continúa con esta actividad completando las siguientes instrucciones.


In [2]:
# RUN THIS CELL TO IMPORT YOUR DATA.

# Save in a variable named `data`.

### YOUR CODE HERE ###

data = pd.read_csv("nba-players.csv", index_col=0)

<details><summary><h4><strong>Pista 1</strong></h4></summary>

La función `read_csv()` de `pandas` te permite leer datos de un archivo csv y cargarlos en un DataFrame.
    
</details>


<details><summary><h4><strong>Pista 2</strong></h4></summary>

Llamar a `read_csv()`, pasar el nombre del archivo csv como una cadena, seguido de `index_col=0` para usar la primera columna del csv como índice en el DataFrame.
    
</details>


## **Paso 2: exploración de datos**


Mostrar las primeras 10 filas de los datos para tener una idea de qué implica.


In [3]:
# Display first 10 rows of data.

### YOUR CODE HERE ###

data.head(10)

,name,gp,min,pts,fgm,fga,fg,3p_made,3pa,3p,...,fta,ft,oreb,dreb,reb,ast,stl,blk,tov,target_5yrs
0,Brandon Ingram,36,27.4,7.4,2.6,7.6,34.7,0.5,2.1,25.0,...,2.3,69.9,0.7,3.4,4.1,1.9,0.4,0.4,1.3,0
1,Andrew Harrison,35,26.9,7.2,2.0,6.7,29.6,0.7,2.8,23.5,...,3.4,76.5,0.5,2.0,2.4,3.7,1.1,0.5,1.6,0
2,JaKarr Sampson,74,15.3,5.2,2.0,4.7,42.2,0.4,1.7,24.4,...,1.3,67.0,0.5,1.7,2.2,1.0,0.5,0.3,1.0,0
3,Malik Sealy,58,11.6,5.7,2.3,5.5,42.6,0.1,0.5,22.6,...,1.3,68.9,1.0,0.9,1.9,0.8,0.6,0.1,1.0,1
4,Matt Geiger,48,11.5,4.5,1.6,3.0,52.4,0.0,0.1,0.0,...,1.9,67.4,1.0,1.5,2.5,0.3,0.3,0.4,0.8,1
5,Tony Bennett,75,11.4,3.7,1.5,3.5,42.3,0.3,1.1,32.5,...,0.5,73.2,0.2,0.7,0.8,1.8,0.4,0.0,0.7,0
6,Don MacLean,62,10.9,6.6,2.5,5.8,43.5,0.0,0.1,50.0,...,1.8,81.1,0.5,1.4,2.0,0.6,0.2,0.1,0.7,1
7,Tracy Murray,48,10.3,5.7,2.3,5.4,41.5,0.4,1.5,30.0,...,0.8,87.5,0.8,0.9,1.7,0.2,0.2,0.1,0.7,1
8,Duane Cooper,65,9.9,2.4,1.0,2.4,39.2,0.1,0.5,23.3,...,0.5,71.4,0.2,0.6,0.8,2.3,0.3,0.0,1.1,0
9,Dave Johnson,42,8.5,3.7,1.4,3.5,38.3,0.1,0.3,21.4,...,1.4,67.8,0.4,0.7,1.1,0.3,0.2,0.0,0.7,0


<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

Hay una función en la biblioteca `pandas` que se puede llamar en un DataFrame para mostrar las primeras n filas, donde n es un número de tu elección.
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Llama a la función `head()` y pasa 10.
</details>


Mostrar el número de filas y el número de columnas para tener una idea de cuántos datos están disponibles para ti.


In [4]:
# Display number of rows, number of columns.

### YOUR CODE HERE ###

data.shape

(1340, 21)

<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

DataFrames en `pandas` tienen un atributo que se puede llamar para obtener el número de filas y columnas como una tupla.
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Puedes llamar al atributo `shape`.
</details>


**Pregunta:** ¿Qué observas acerca del número de filas y el número de columnas en los datos?
- Hay 1,340 filas y 21 columnas en los datos.


Ahora, muestra todos los nombres de columna para tener una idea de los tipos de metadatos disponibles sobre cada jugador. Usa la propiedad columns en pandas.


In [5]:
# Display all column names.

### YOUR CODE HERE ###

data.columns

Index(['name', 'gp', 'min', 'pts', 'fgm', 'fga', 'fg', '3p_made', '3pa', '3p',
       'ftm', 'fta', 'ft', 'oreb', 'dreb', 'reb', 'ast', 'stl', 'blk', 'tov',
       'target_5yrs'],
      dtype='object')

La siguiente tabla proporciona una descripción de los datos en cada columna. Estos metadatos provienen de la fuente de datos, que se encuentra en la sección de referencias de este laboratorio.

<center>

|Nombre de la columna|Descripción de la columna|
|:---|:-------|
|`name`|Nombre del jugador de la NBA|
|`gp`|Número de partidos jugados|
|`min`|Número de minutos jugados por partido|
|`pts`|Número promedio de puntos por partido|
|`fgm`|Número promedio de tiros de campo anotados por partido|
|`fga`|Número promedio de intentos de tiros de campo por partido|
|`fg`|Porcentaje promedio de tiros de campo anotados por partido|
|`3p_made`|Número promedio de tiros de tres puntos anotados por partido|
|`3pa`|Número promedio de intentos de tiros de tres puntos por partido|
|`3p`|Porcentaje promedio de tiros de tres puntos anotados por partido|
|`ftm`|Número promedio de tiros libres anotados por partido|
|`fta`|Número promedio de intentos de tiros libres por partido|
|`ft`|Porcentaje promedio de tiros libres anotados por partido|
|`oreb`|Número promedio de rebotes ofensivos por partido|
|`dreb`|Número promedio de rebotes defensivos por partido|
|`reb`|Número promedio de rebotes por partido|
|`ast`|Número promedio de asistencias por partido|
|`stl`|Número promedio de robos por partido|
|`blk`|Número promedio de bloqueos por partido|
|`tov`|Número promedio de pérdidas por partido|
|`target_5yrs`|1 si la duración de la carrera >= 5 años, 0 en caso contrario|

</center>


A continuación, muestre un resumen de los datos para obtener información adicional sobre el DataFrame, incluyendo los tipos de datos en las columnas.


In [6]:
# Use .info() to display a summary of the DataFrame.

### YOUR CODE HERE ###

data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1340 non-null   object 
 1   gp           1340 non-null   int64  
 2   min          1340 non-null   float64
 3   pts          1340 non-null   float64
 4   fgm          1340 non-null   float64
 5   fga          1340 non-null   float64
 6   fg           1340 non-null   float64
 7   3p_made      1340 non-null   float64
 8   3pa          1340 non-null   float64
 9   3p           1340 non-null   float64
 10  ftm          1340 non-null   float64
 11  fta          1340 non-null   float64
 12  ft           1340 non-null   float64
 13  oreb         1340 non-null   float64
 14  dreb         1340 non-null   float64
 15  reb          1340 non-null   float64
 16  ast          1340 non-null   float64
 17  stl          1340 non-null   float64
 18  blk          1340 non-null   float64
 19  tov   

**Pregunta:** Basándose en las tablas anteriores, ¿cuáles columnas son numéricas y cuáles columnas son categóricas?
- Basándose en las tablas anteriores, la columna `name` es categórica, y el resto de las columnas son numéricas.


### Verificar valores faltantes


Ahora, revisa los datos para determinar si contienen valores faltantes. Comienza mostrando el número de valores faltantes en cada columna. Después, usa isna() para verificar si cada valor en los datos falta. Finalmente, usa sum() para agregar el número de valores faltantes por columna.


In [7]:
# Display the number of missing values in each column.
# Check whether each value is missing.
#Aggregate the number of missing values per column.

### YOUR CODE HERE ###

data.isna().sum()

name           0
gp             0
min            0
pts            0
fgm            0
fga            0
fg             0
3p_made        0
3pa            0
3p             0
ftm            0
fta            0
ft             0
oreb           0
dreb           0
reb            0
ast            0
stl            0
blk            0
tov            0
target_5yrs    0
dtype: int64

**Pregunta:** ¿Qué observas acerca de los valores faltantes en las columnas?  
Todas las columnas en este conjunto de datos tienen 0 valores faltantes.


**Pregunta:** ¿Por qué es importante verificar la presencia de valores faltantes?  
Verificar la presencia de valores faltantes es un paso importante en la exploración de datos. Los valores faltantes no son particularmente útiles, por lo que es importante manejarlos limpiando los datos.


## **Paso 3: Pruebas estadísticas**


A continuación, utilice una técnica estadística para verificar el equilibrio de clases en los datos. Para entender qué tan equilibrado está el conjunto de datos en términos de clase, muestre el porcentaje de valores que pertenecen a cada clase en la columna objetivo. En este contexto, la clase 1 indica una duración de carrera en la NBA de al menos cinco años, mientras que la clase 0 indica una duración de carrera en la NBA de menos de cinco años.


In [8]:
# Display percentage (%) of values for each class (1, 0) represented in the target column of this dataset.

### YOUR CODE HERE ###

data["target_5yrs"].value_counts(normalize=True)*100

1    62.014925
0    37.985075
Name: target_5yrs, dtype: float64

<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

En `pandas`, `value_counts(normalize=True)` puede ser utilizado para calcular la frecuencia de cada valor distinto en una columna específica de un DataFrame.  
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Después de `value_counts(normalize=True)`, multiplicar por `100` convierte las frecuencias en porcentajes (%).
</details>


**Pregunta:** ¿Qué observas sobre el equilibrio de clases en la columna objetivo?
- Aproximadamente el 62% de los valores en la columna objetivo pertenecen a la clase 1, y aproximadamente el 38% de los valores pertenecen a la clase 0. En otras palabras, aproximadamente el 62% de los jugadores representados por estos datos tienen una duración de carrera en la NBA de al menos cinco años, y aproximadamente el 38% no.
- El conjunto de datos no está perfectamente equilibrado, pero una división exacta 50-50 es una ocurrencia rara en los conjuntos de datos, y una división 62-38 no está demasiado desequilibrada. Sin embargo, si la clase mayoritaria representara el 90% o más del conjunto de datos, entonces eso sería motivo de preocupación, y sería prudente abordar ese problema mediante técnicas como el sobremuestreo y el submuestreo.


**Pregunta:** ¿Por qué es importante verificar el equilibrio de clases?


Si hay mucha más representación de una clase que de otra, entonces el modelo puede estar sesgado hacia la clase mayoritaria. Cuando esto sucede, las predicciones pueden ser inexactas.


## **Paso 4: Resultados y evaluación**


Ahora, realiza ingeniería de características, con el objetivo de identificar y crear características que sirvan como predictores útiles para la variable objetivo, `target_5yrs`.


### Selección de Características


La siguiente tabla contiene descripciones de los datos en cada columna:

<center>

|Nombre de la columna|Descripción de la columna|
|:---|:-------|
|`name`|Nombre del jugador de la NBA|
|`gp`|Número de partidos jugados|
|`min`|Número de minutos jugados|
|`pts`|Promedio de puntos por partido|
|`fgm`|Promedio de canastas de campo hechas por partido|
|`fga`|Promedio de intentos de canasta de campo por partido|
|`fg`|Porcentaje promedio de canastas de campo hechas por partido|
|`3p_made`|Promedio de canastas de tres puntos hechas por partido|
|`3pa`|Promedio de intentos de canasta de tres puntos por partido|
|`3p`|Porcentaje promedio de canastas de tres puntos hechas por partido|
|`ftm`|Promedio de tiros libres hechos por partido|
|`fta`|Promedio de intentos de tiros libres por partido|
|`ft`|Porcentaje promedio de tiros libres hechos por partido|
|`oreb`|Promedio de rebotes ofensivos por partido|
|`dreb`|Promedio de rebotes defensivos por partido|
|`reb`|Promedio de rebotes por partido|
|`ast`|Promedio de asistencias por partido|
|`stl`|Promedio de robos por partido|
|`blk`|Promedio de bloqueos por partido|
|`tov`|Promedio de pérdidas por partido|
|`target_5yrs`|1 si la duración de la carrera >= 5 años, 0 en caso contrario|

</center>


Pregunta: ¿Qué columnas seleccionarías y cuáles evitarías seleccionar como características, y por qué? Ten en cuenta que el objetivo es identificar características que sirvan como predictores útiles para la variable objetivo, `target_5yrs`.

- Deberías evitar seleccionar la columna `name` como característica. El nombre de un jugador no ayuda a determinar la duración de su carrera. Además, puede no ser ético o justo predecir la duración de la carrera de un jugador basándose en un nombre.
- El número de partidos en los que un jugador ha participado puede no ser tan importante para determinar la duración de su carrera como la cantidad de puntos que ha obtenido. Aunque podrías decir que alguien que ha jugado en más partidos puede tener más práctica y experiencia, los puntos que obtiene durante los partidos en los que jugó hablarían más de su rendimiento como jugador. Esto, a su vez, influiría en la duración de su carrera. Entonces, la columna `gp` por sí sola puede no ser una característica útil. Sin embargo, `gp` y `pts` podrían combinarse para obtener el total de puntos obtenidos en todos los partidos, y ese resultado podría ser una característica útil. Ese enfoque puede implementarse más adelante en el proceso de ingeniería de características—en la extracción de características.
- Si se extraerá como característica el número de puntos obtenidos en los partidos, entonces eso podría combinarse con el número de minutos jugados en los partidos (`min * gp`) para extraer otra característica. Esto podría ser una medida de eficiencia de los jugadores y podría ayudar a predecir la duración de la carrera de los jugadores. `min` por sí solo puede no ser útil como característica por la misma razón que `gp`.
- Hay tres columnas diferentes que dan información sobre tiros de campo. El porcentaje de tiros de campo que un jugador encesta (`fg`) dice más sobre su rendimiento que el número de tiros de campo que encesta (`fgm`) o el número de tiros de campo que intenta (`fga`). El porcentaje da más contexto, ya que tiene en cuenta tanto cuántos tiros de campo un jugador logró con éxito como cuántos intentó en total. Esto permite una comparación más significativa entre jugadores. La misma lógica se aplica al porcentaje de tiros de tres puntos encestados, así como al porcentaje de tiros libres encestados.
- Hay columnas para el número de rebotes ofensivos (`oreb`), el número de rebotes defensivos (`dreb`) y el número total de rebotes (`reb`). Debido a que el número total de rebotes ya debería incorporar tanto los rebotes ofensivos como los defensivos, tendría sentido usar el total como una característica.
- El número de asistencias (`ast`), robos (`stl`), bloqueos (`blk`) y pérdidas (`tov`) también proporcionan información sobre qué tan bien están rindiendo los jugadores en los partidos, y por lo tanto, podrían ser útiles para predecir cuánto tiempo permanecen los jugadores en la liga.

Por lo tanto, en esta etapa del proceso de ingeniería de características, sería más efectivo seleccionar las siguientes columnas:

`gp`, `min`, `pts`, `fg`, `3p`, `ft`, `reb`, `ast`, `stl`, `blk`, `tov`.


A continuación, selecciona las columnas con las que deseas proceder. Asegúrate de incluir la columna objetivo, `target_5yrs`. Muestra las primeras filas para confirmar que son como se espera.


In [9]:
# Select the columns to proceed with and save the DataFrame in new variable `selected_data`.
# Include the target column, `target_5yrs`.

### YOUR CODE HERE ###

selected_data = data[["gp", "min", "pts", "fg", "3p", "ft", "reb", "ast", "stl", "blk", "tov", "target_5yrs"]]


# Display the first few rows.

### YOUR CODE HERE ###


selected_data.head()

,gp,min,pts,fg,3p,ft,reb,ast,stl,blk,tov,target_5yrs
0,36,27.4,7.4,34.7,25.0,69.9,4.1,1.9,0.4,0.4,1.3,0
1,35,26.9,7.2,29.6,23.5,76.5,2.4,3.7,1.1,0.5,1.6,0
2,74,15.3,5.2,42.2,24.4,67.0,2.2,1.0,0.5,0.3,1.0,0
3,58,11.6,5.7,42.6,22.6,68.9,1.9,0.8,0.6,0.1,1.0,1
4,48,11.5,4.5,52.4,0.0,67.4,2.5,0.3,0.3,0.4,0.8,1


<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

Consulte los materiales sobre selección de características y selección de un subconjunto de un DataFrame.
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Utilice dos pares de corchetes cuadrados, y coloque los nombres de las columnas que desea seleccionar dentro de los corchetes más internos. 

</details>


<details>
<summary><h4><strong>Pista 3</strong></h4></summary>

Hay una función en `pandas` que se puede usar para mostrar las primeras filas de un DataFrame. Asegúrate de especificar los nombres de las columnas con la ortografía que coincida con lo que hay en los datos. Usa comillas para representar cada nombre de columna como una cadena.
</details>


### Transformación de características


Un aspecto importante de la transformación de características es la codificación de características. Si hay columnas categóricas que deseas usar como características, esas columnas deben ser transformadas para ser numéricas. Esta técnica también se conoce como codificación de características.


**Pregunta:** ¿Por qué es importante considerar la transformación de características? ¿Hay alguna transformación necesaria para las características que deseas usar?
- Muchos tipos de modelos están diseñados de manera que requieren que los datos que ingresan sean numéricos. Entonces, transformar características categóricas en características numéricas es un paso importante.
- En este conjunto de datos en particular, `name` es la única columna categórica y las otras columnas son numéricas (discutido en la respuesta ejemplar a la Pregunta 2). Dado que `name` no está seleccionado como una característica, todas las características que se seleccionan en este momento ya son numéricas y no requieren transformación.


### Extracción de características


Mostrar las primeras filas que contienen descripciones de los datos para referencia. La tabla es la siguiente:

<center>

|Nombre de la columna|Descripción de la columna|
|:---|:-------|
|`name`|Nombre del jugador de la NBA|
|`gp`|Número de partidos jugados|
|`min`|Número de minutos jugados por partido|
|`pts`|Número promedio de puntos por partido|
|`fgm`|Número promedio de tiros de campo anotados por partido|
|`fga`|Número promedio de intentos de tiros de campo por partido|
|`fg`|Porcentaje promedio de tiros de campo anotados por partido|
|`3p_made`|Número promedio de tiros de tres puntos anotados por partido|
|`3pa`|Número promedio de intentos de tiros de tres puntos por partido|
|`3p`|Porcentaje promedio de tiros de tres puntos anotados por partido|
|`ftm`|Número promedio de tiros libres anotados por partido|
|`fta`|Número promedio de intentos de tiros libres por partido|
|`ft`|Porcentaje promedio de tiros libres anotados por partido|
|`oreb`|Número promedio de rebotes ofensivos por partido|
|`dreb`|Número promedio de rebotes defensivos por partido|
|`reb`|Número promedio de rebotes por partido|
|`ast`|Número promedio de asistencias por partido|
|`stl`|Número promedio de robos por partido|
|`blk`|Número promedio de bloqueos por partido|
|`tov`|Número promedio de pérdidas por partido|
|`target_5yrs`|1 si la duración de la carrera >= 5 años, 0 en caso contrario|

</center>


In [10]:
# Display the first few rows of `selected_data` for reference.

### YOUR CODE HERE ###

selected_data.head()

,gp,min,pts,fg,3p,ft,reb,ast,stl,blk,tov,target_5yrs
0,36,27.4,7.4,34.7,25.0,69.9,4.1,1.9,0.4,0.4,1.3,0
1,35,26.9,7.2,29.6,23.5,76.5,2.4,3.7,1.1,0.5,1.6,0
2,74,15.3,5.2,42.2,24.4,67.0,2.2,1.0,0.5,0.3,1.0,0
3,58,11.6,5.7,42.6,22.6,68.9,1.9,0.8,0.6,0.1,1.0,1
4,48,11.5,4.5,52.4,0.0,67.4,2.5,0.3,0.3,0.4,0.8,1


Pregunta: ¿Qué columnas se prestan para la extracción de características?

- Las columnas `gp`, `pts`, `min` se prestan para la extracción de características.
  - `gp` representa el número total de partidos que un jugador ha jugado, y `pts` representa el promedio de puntos que el jugador ha obtenido por partido. Podría ser útil combinar estas columnas para obtener el total de puntos que el jugador ha obtenido en los partidos y usar el resultado como una nueva característica, que podría añadirse en una nueva columna llamada `total_points`. Los puntos totales obtenidos por un jugador pueden reflejar su rendimiento y moldear la longevidad de su carrera.
  - La columna `min` representa el promedio de minutos jugados por partido. `total_points` podría combinarse con `min` y `gp` para extraer una nueva característica: puntos obtenidos por minuto. Esto puede considerarse una medida de eficiencia del jugador, que podría moldear la duración de la carrera. Esta característica puede añadirse en una columna llamada `efficiency`.


Extrae dos características que crees que ayudarían a predecir `target_5yrs`. Luego, crea una nueva variable llamada 'extracted_data' que contenga las características de 'selected_data', así como las características que se están extrayendo.


In [11]:
# Extract two features that would help predict target_5yrs.
# Create a new variable named `extracted_data`.

### YOUR CODE HERE ### 

# Make a copy of `selected_data` 
extracted_data = selected_data.copy()

# Add a new column named `total_points`; 
# Calculate total points earned by multiplying the number of games played by the average number of points earned per game
extracted_data["total_points"] = extracted_data["gp"] * extracted_data["pts"]

# Add a new column named `efficiency`. Calculate efficiency by dividing the total points earned by the total number 
# of minutes played, which yields points per minute. (Note that `min` represents avg. minutes per game.)
extracted_data["efficiency"] = extracted_data["total_points"] / (extracted_data["min"] * extracted_data["gp"])

# Display the first few rows of `extracted_data` to confirm that the new columns were added.
extracted_data.head()

,gp,min,pts,fg,3p,ft,reb,ast,stl,blk,tov,target_5yrs,total_points,efficiency
0,36,27.4,7.4,34.7,25.0,69.9,4.1,1.9,0.4,0.4,1.3,0,266.4,0.270073
1,35,26.9,7.2,29.6,23.5,76.5,2.4,3.7,1.1,0.5,1.6,0,252.0,0.267658
2,74,15.3,5.2,42.2,24.4,67.0,2.2,1.0,0.5,0.3,1.0,0,384.8,0.339869
3,58,11.6,5.7,42.6,22.6,68.9,1.9,0.8,0.6,0.1,1.0,1,330.6,0.491379
4,48,11.5,4.5,52.4,0.0,67.4,2.5,0.3,0.3,0.4,0.8,1,216.0,0.391304


<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

Consulte los materiales sobre extracción de características.
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Utilice la función `copy()` para hacer una copia de un DataFrame. Para acceder a una columna específica de un DataFrame, utilice un par de corchetes cuadrados y coloque el nombre de la columna como una cadena dentro de los corchetes.

</details>


<details>
<summary><h4><strong>Pista 3</strong></h4></summary>

Utiliza un par de corchetes para crear una nueva columna en un DataFrame. Las columnas en los DataFrames son objetos de series, que soportan operaciones elemento a elemento como multiplicación y división. Asegúrate de que los nombres de las columnas referenciados en tu código coincidan con la ortografía de lo que hay en el DataFrame.
</details>


Ahora, para prepararse para el modelo de Naive Bayes que construirás en un laboratorio posterior, limpia los datos extraídos y asegúrate de que sean concisos. Naive Bayes implica una suposición de que las características son independientes entre sí dado la clase. Para cumplir con ese criterio, si ciertas características se agregan para obtener nuevas características, puede ser necesario eliminar esas características originales. Por lo tanto, elimina las columnas que se usaron para extraer nuevas características.

**Nota:** Hay otros tipos de modelos que no implican suposiciones de independencia, por lo que esto no sería necesario en esos casos. De hecho, mantener las características originales puede ser beneficioso.


In [12]:
# Remove any columns from `extracted_data` that are no longer needed.

### YOUR CODE HERE ###

# Remove `gp`, `pts`, and `min` from `extracted_data`.
extracted_data = extracted_data.drop(columns=["gp", "pts", "min"])

# Display the first few rows of `extracted_data` to ensure that column drops took place.

### YOUR CODE HERE ###

extracted_data.head()

,fg,3p,ft,reb,ast,stl,blk,tov,target_5yrs,total_points,efficiency
0,34.7,25.0,69.9,4.1,1.9,0.4,0.4,1.3,0,266.4,0.270073
1,29.6,23.5,76.5,2.4,3.7,1.1,0.5,1.6,0,252.0,0.267658
2,42.2,24.4,67.0,2.2,1.0,0.5,0.3,1.0,0,384.8,0.339869
3,42.6,22.6,68.9,1.9,0.8,0.6,0.1,1.0,1,330.6,0.491379
4,52.4,0.0,67.4,2.5,0.3,0.3,0.4,0.8,1,216.0,0.391304


<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

Consulte los materiales sobre extracción de características.
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Hay funciones en la biblioteca `pandas` que eliminan columnas específicas de un DataFrame y que muestran las primeras filas de un DataFrame.
</details>


<details>
<summary><h4><strong>Pista 3</strong></h4></summary>

Utiliza la función `drop()` y pasa una lista con los nombres de las columnas que deseas eliminar. Por defecto, llamar a esta función resultará en un nuevo DataFrame que refleja los cambios realizados. El DataFrame original no se altera automáticamente. Puedes reasignar `extracted_data` al resultado, para actualizarlo.

Utiliza la función `head()` para mostrar las primeras filas de un DataFrame.
</details>


A continuación, exporta los datos extraídos como un nuevo archivo .csv. Utilizarás esto en un laboratorio posterior.


In [13]:
# Export the extracted data.

### YOUR CODE HERE ###

extracted_data.to_csv("extracted_nba_players_data.csv", index=0)

<details>
<summary><h4><strong>Pista 1</strong></h4></summary>

Hay una función en la biblioteca `pandas` que exporta un DataFrame como un archivo .csv.
</details>


<details>
<summary><h4><strong>Pista 2</strong></h4></summary>

Utilice la función `to_csv()` para exportar el DataFrame como un archivo .csv.
</details>


<details>
<summary><h4><strong>Pista 3</strong></h4></summary>

Llama a la función `to_csv()` en `extracted_data`, y pasa el nombre que deseas darle al archivo .csv resultante. Especifica el nombre del archivo como una cadena y en el nombre del archivo. Asegúrate de incluir `.csv` como la extensión del archivo. 
    
    Además, pasa el parámetro `index` establecido en `0`, para que cuando ocurra la exportación, los índices de fila del DataFrame no se traten como una columna adicional en el archivo resultante. 
</details>


## **Consideraciones**


**¿Cuáles son algunos aspectos clave que aprendiste durante este laboratorio?**
- Es importante verificar el equilibrio de clases en un conjunto de datos, particularmente en el contexto de ingeniería de características y modelado predictivo. Si la columna objetivo en un conjunto de datos tiene más del 90% de sus valores pertenecientes a una clase, se recomienda redistribuir los datos; de lo contrario, una vez que un modelo se entrena con los datos desequilibrados y se hacen predicciones, las predicciones pueden estar sesgadas. 
- La selección de características implica elegir características que ayuden a predecir la variable objetivo y eliminar columnas que pueden no ser útiles para la predicción. En este proceso, y a lo largo de la ingeniería de características, es importante hacer consideraciones éticas.  
- La transformación de características implica transformar las características para que sean más útiles para futuros propósitos de modelado, lo cual incluye codificar características categóricas para convertirlas en características numéricas. 
- La extracción de características implica combinar columnas existentes de manera significativa para construir nuevas características que ayudarían a mejorar la predicción. 

**¿Qué resumen proporcionarías a las partes interesadas? Considera atributos clave para compartir de los datos, así como planes de proyectos futuros.**
- Los siguientes atributos sobre el rendimiento de los jugadores podrían ayudar a predecir la duración de su carrera en la NBA y deberían incluirse en una presentación para las partes interesadas: tiros de campo, tiros de campo de tres puntos, tiros libres, rebotes, asistencias, robos, bloqueos, pérdidas, puntos totales y eficiencia como puntos por minuto. 
- Sería importante explicar que estos atributos, junto con un conjunto de datos relevante, se utilizarán en la siguiente etapa del proyecto. En ese momento, se construirá un modelo para predecir la duración de la carrera de un jugador. Los conocimientos adquiridos se compartirán con las partes interesadas una vez que el proyecto esté completo. Las partes interesadas también agradecerían que se les proporcionara una línea de tiempo y entregables clave que puedan esperar recibir.


¡Felicidades! Has completado este laboratorio. Sin embargo, es posible que no notes una marca de verificación verde junto a este elemento en la plataforma de Coursera. Por favor, continúa con tu progreso independientemente de la marca de verificación. Solo haz clic en el icono de "guardar" en la parte superior de este cuaderno para asegurarte de que tu trabajo ha sido registrado
